### ISSUE TITLE
task_namespace lost during inheritance

### ISSUE THREAD

**Issue by riga** [spotify/luigi#1950]
Hey guys!

When inheriting from a task class that has a `task_namespace`, imho it would be pythonic that the inheriting class has the same `task_namespace` (if not set otherwise). However, this does not seem to be the case. Is this behavior intended by design?

### Possible solution

Due to the way the task register meta class implements the `_default_namespace` fallback, the proper namespace is only set when it's contained in the `class_dict`: ([link](https://github.com/spotify/luigi/blob/master/luigi/task_register.py#L67))

```python
class Register(abc.ABCMeta):

    def __new__(metacls, classname, bases, classdict):
        if "task_namespace" not in classdict:
            classdict["task_namespace"] = metacls._default_namespace

        cls = super(Register, metacls).__new__(metacls, classname, bases, classdict)
        metacls._reg.append(cls)
```

The fix would be quite simple:

```python
class Register(abc.ABCMeta):

    def __new__(metacls, classname, bases, classdict):
        cls = super(Register, metacls).__new__(metacls, classname, bases, classdict)

        if not getattr(cls, "task_namespace", None):
            cls.task_namespace = metacls._default_namespace

        metacls._reg.append(cls)
```

This implies a minor change though: the inherited namespace would be favored over the default one, except when the inherited one is falsy.

Any thoughts?

**Comment by Tarrasch** [comment]
I'm all for this change. Pretty sure this is an old old mistake lingering around. Hopefully not too many rely on this unpythonic behavior.

**Comment by Tarrasch** [comment]
I reopen this issue. Because I've found the recent PR to be insufficient. The good news is that I already know a sort of a solution and I'm willing to work on it myself. Read on for details:

## Current problem since #1953 

First, the implemented solution is okay, but at least my code started to break immediately. I Had this pattern.

```
$ cat file_a.py
luigi_namespace(__module__)

class GenericBaseTask(luigi.Task)
    pass

$ cat file_b.py
import file_a
luigi_namespace(__module__)

class SpecificTask(file_a.GenericBaseTask)
    pass
```

Before #1953, I would have `file_b.SpecificTask` which I find logical. I mean it's created there and there's this `luigi_namespace(__module__)` part in that file as well.

## Proposed solution

I'll rather just implement this. I'll try to do it today. :)

**Comment by Tarrasch** [comment]
Please see #1965 @riga, it supports both the case you originally wanted fixed and the case I described above. Please review it. :)

**Comment by Tarrasch** [comment]
Closing again. This got fixed in #1965 again.
